### PR TITLE
[darknet] fix opencv features, add compatibility with VS16.9

### LINKED ISSUE #16011: [darknet] build failure
**Host Environment**
- OS: [e.g. Windows/Linux etc...]
- Compiler: revision
Windows
**To Reproduce**
`./vcpkg install xxxx`

**Failure logs** 
```
-Cut and past the appropriate build messages from the console output.
-Please attach any additional failure logs mentioned in the console output.
Starting package 1/1: darknet:x64-windows
Building package darknet[core,cuda,cudnn,full,opencv-cuda,weights,weights-train]:x64-windows...
-- Using cached C:/Users/Kun/vcpkg/downloads/AlexeyAB-darknet-00d578e327c22638ea12e73c4efb74c798c08de5.tar.gz
-- Cleaning sources at C:/Users/Kun/vcpkg/buildtrees/darknet/src/c798c08de5-d29893808b.clean. Use --editable to skip cleaning for the packages you specify.
-- Extracting source C:/Users/Kun/vcpkg/downloads/AlexeyAB-darknet-00d578e327c22638ea12e73c4efb74c798c08de5.tar.gz
-- Using source at C:/Users/Kun/vcpkg/buildtrees/darknet/src/c798c08de5-d29893808b.clean
-- Downloading https://github.com/AlexeyAB/darknet/releases/download/darknet_yolo_v4_pre/yolov4-tiny.weights -> darknet-cache/yolov4-tiny.weights...
CMake Error at scripts/cmake/vcpkg_download_distfile.cmake:105 (message):


  File does not have expected hash:

          File path: [ C:/Users/Kun/vcpkg/downloads/temp/darknet-cache/yolov4-tiny.weights ]
      Expected hash: [ 804ca2ab8e3699d31c95bf773d22f901f186703487c7945f30dc2dbb808094793362cb6f5da5cd0b4b83f820c8565a3cba22fafa069ee6ca2a925677137d95f4 ]
        Actual hash: [ 7d4d9fe150f9fe3ea7d2310f1445fe983b31fbf06d301c70ecfe00e8559e6f1bf940198c2dd55db772238f23ea0092fb6553558e5414f3ee173b8b28e53c5b54 ]

  The file may have been corrupted in transit.  This can be caused by
  proxies.  If you use a proxy, please set the HTTPS_PROXY and HTTP_PROXY
  environment variables to
  "https://user:password@your-proxy-ip-address:port/".



Call Stack (most recent call first):
  scripts/cmake/vcpkg_download_distfile.cmake:195 (test_hash)
  ports/darknet/portfile.cmake:28 (vcpkg_download_distfile)
  scripts/ports.cmake:128 (include)


Error: Building package darknet:x64-windows failed with: BUILD_FAILED
Please ensure you're using the latest portfiles with `.\vcpkg update`, then
submit an issue at https://github.com/Microsoft/vcpkg/issues including:
  Package: darknet:x64-windows
  Vcpkg version: 2021-01-13-unknownhash

Additionally, attach any relevant sections from the log files above.
```
**Additional context**
Add any other context about the problem here, such as what you have already tried to resolve the issue.

### PR DESCRIPTION
**Pull request by cenit**
as per title, otherwise opencv features would disable each others (only last one surviving)

also, VS16.9 introduced a big compatibility breaking which disrupted some projects, including darknet https://github.com/AlexeyAB/darknet/issues/7533

### CODE REVIEW

**Comment by cenit** [comment]
@JackBoosY PR is ready for merge, sorry for the very long delay
@ras0219-msft I dropped the weight and weight-train features and instead opted for a ps1 script for users to download weights manually afterwards

**Comment by cenit** [comment]
@strega-nil as promised (back in february, i think 😢 ) now darknet shares the manifest files with the real project itself 👍🏻

File: ports/darknet/vcpkg.json [hunk 1]
```diff
@@ -0,0 +1,118 @@
+{
+  "name": "darknet",
+  "version-string": "2021-04-16",
```
Comment (JackBoosY) [inline ports/darknet/vcpkg.json:3]:
```suggestion
  "version-date": "2021-04-16",
```

**Review by JackBoosY — APPROVED** [review]

File: versions/d-/darknet.json [hunk 2]
```diff
@@ -1,5 +1,10 @@
 {
   "versions": [
+    {
+      "git-tree": "f4828e2186d8f9d42a1679c1d01c85f7f43bc462",
+      "version-string": "2021-04-16",
```
Comment (JackBoosY) [inline versions/d-/darknet.json:5]:
```suggestion
      "version-date": "2021-04-16",
```

**Review by JackBoosY — CHANGES_REQUESTED** [review]

File: ports/darknet/portfile.cmake [hunk 3]
```diff
@@ -1,95 +1,35 @@
...
-  REF 00d578e327c22638ea12e73c4efb74c798c08de5
-  SHA512 ef2d46fab670759e9c22d0233b60192bfe47669e29d2ec1e020a77dfaf09894a93160c11de070bc39d86109dd2a37ca7172fbb081809b1ea2783207a6e385b2c
+  REF d7c37b5616038b9283a37e043cfeadd26b182da1
+  SHA512 6d237f2049111c62be9a9312478b6debbac9a92101aa1d3ff3eadcf1e9eecb056f1ee80b7f7899e8c9de1432015245019d07ddc089c76e732b02ba91aae8f7f0
   HEAD_REF master
 )
 
 vcpkg_check_features(OUT_FEATURE_OPTIONS FEATURE_OPTIONS
+    FEATURES
     cuda ENABLE_CUDA
```
Comment (NancyLi1013) [inline ports/darknet/portfile.cmake:11]:
```suggestion
        cuda ENABLE_CUDA
```
Please also update the following line:
`cudnn ENABLE_CUDNN`

File: ports/darknet/portfile.cmake [hunk 4]
```diff
@@ -1,95 +1,35 @@
...
-    opencv3-cuda ENABLE_OPENCV
     cudnn ENABLE_CUDNN
 )
 
+#do not move following features to vcpkg_check_features because they break themselves: one off will turn off the others even if true
+set(ENABLE_OPENCV FALSE)
+if ("opencv-base" IN_LIST FEATURES OR "opencv2-base" IN_LIST FEATURES OR "opencv3-base" IN_LIST FEATURES)
+  set(ENABLE_OPENCV TRUE)
+endif()
+set(ENABLE_OPENCV FALSE)
```
Comment (NancyLi1013) [inline ports/darknet/portfile.cmake:20]: Why set `ENABLE_OPENCV ` as `false `again here?

**Review by NancyLi1013 — CHANGES_REQUESTED** [review]

File: ports/darknet/portfile.cmake (hunk 4 above)
Comment (cenit) [reply ports/darknet/portfile.cmake:20]: my mistake! thanks for noticing

**Comment by cenit** [comment]
thanks for the review @NancyLi1013 @JackBoosY

**Review by JackBoosY — APPROVED** [review]

**Comment by JackBoosY** [comment]
LGTM in my side.

File: versions/baseline.json [hunk 5]
```diff
@@ -1557,8 +1557,8 @@
       "port-version": 0
     },
     "darknet": {
-      "baseline": "2021-01-21",
-      "port-version": 0
+      "baseline": "2021-04-16",
+      "port-version": 1
```
Comment (NancyLi1013) [inline versions/baseline.json:1561]:
```suggestion
      "port-version": 0
```

File: versions/d-/darknet.json [hunk 6]
```diff
@@ -1,5 +1,10 @@
 {
   "versions": [
+    {
+      "git-tree": "d231e3fa1427d0744b88ab6a60d34870132e1f83",
+      "version-date": "2021-04-16",
+      "port-version": 1
```
Comment (NancyLi1013) [inline versions/d-/darknet.json:6]:
```suggestion
      "port-version": 0
```

**Comment by NancyLi1013** [comment]
Except for the above suggestion, LGTM now, thanks for your efforts @cenit

**Comment by cenit** [comment]
moved back to draft because a suspect bug has just been found upstream, a quick fix should come
https://github.com/AlexeyAB/darknet/pull/7610

**Comment by cenit** [comment]
bug fix done upstream

bonus: darknet should work also on windows arm architectures :)

**Comment by cenit** [comment]
PR is ready for review again (and merge if ok)
@JackBoosY @NancyLi1013

**Comment by cenit** [comment]
@JackBoosY @NancyLi1013 PR is ready for review again (and merge if ok)

**Review by JackBoosY — APPROVED** [review]

**Review by NancyLi1013 — APPROVED** [review]

**Comment by NancyLi1013** [comment]
Needs to test features.

**Comment by cenit** [comment]
ok thanks. Anything I might do to speed up things?
Without this PR darknet is completely broken with VS 16.9, so I hope for a quick turnaround

**Comment by JackBoosY** [comment]
@cenit We will make this PR high priority.

File: ports/darknet/vcpkg.json [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+        }
+      ]
+    },
+    "opencv2-base": {
+      "description": "Build darknet with support for OpenCV2",
+      "dependencies": [
+        {
+          "name": "opencv2",
+          "features": [
+            "contrib",
```
Comment (NancyLi1013) [inline ports/darknet/vcpkg.json:70]: This feature seems not in `opencv2`. Could you please help check if you missed this feature for `opencv2`?

File: ports/darknet/vcpkg.json [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+        }
+      ]
+    },
+    "opencv2-cuda": {
+      "description": "Build darknet with support for CUDA-enabled OpenCV2",
+      "dependencies": [
+        {
+          "name": "opencv2",
+          "features": [
+            "contrib",
```
Comment (NancyLi1013) [inline ports/darknet/vcpkg.json:82]: The same as above.

**Review by NancyLi1013 — CHANGES_REQUESTED** [review]

File: ports/darknet/vcpkg.json (hunk 7 above)
Comment (cenit) [reply ports/darknet/vcpkg.json]: no sorry my mistake in converting control file to manifest

File: ports/darknet/vcpkg.json (hunk 8 above)
Comment (cenit) [reply ports/darknet/vcpkg.json]: same as above

**Comment by NancyLi1013** [comment]
`Opencv2 `failed with the following errors in feature testing:

```
nvcc fatal   : Unsupported gpu architecture 'compute_30'
CMake Error at cuda_compile_1_generated_matrix_operations.cu.obj.Debug.cmake:220 (message):
  Error generating
  F:/15942/vcpkg/buildtrees/opencv2/x64-windows-dbg/modules/core/CMakeFiles/cuda_compile_1.dir/__/dynamicuda/src/cuda/./cuda_compile_1_generated_matrix_operations.cu.obj


ninja: build stopped: subcommand failed.
```

**Comment by cenit** [comment]
ok, let me take a look at it

**Comment by NancyLi1013** [comment]
I used the command:

`.\vcpkg.exe install darknet[cuda,opencv2-base,opencv2-cuda]:x64-windows`

Note: Since `cudnn `cannot be installed successfully on my machine (CUDA 11.3), I didn't test these features that depends on `cudnn`.

**Comment by cenit** [comment]
> I used the command:
> 
> `.\vcpkg.exe install darknet[cuda,opencv2-base,opencv2-cuda]:x64-windows`
> 
> Note: Since `cudnn `cannot be installed successfully on my machine (CUDA 11.3), I didn't test these features that depends on `cudnn`.

I cannot reproduce your issue locally (but I have CUDA 11.0).

about cudnn: it's a known problem for now, you need to install cudnn manually as the port says, from https://developer.nvidia.com/cudnn

**Comment by NancyLi1013** [comment]
Can you test these features in this PR?

Sorry, I was blocked by the problem caused by `opencv2 `.

**Comment by cenit** [comment]
```
Starting package 3/6: opencv2:x64-windows
Building package opencv2[core,cuda,eigen,ffmpeg,jpeg,png,tiff]:x64-windows...
-- Using cached F:/vcpkg_cenit/downloads/opencv-opencv-2.4.13.7.tar.gz
-- Extracting source F:/vcpkg_cenit/downloads/opencv-opencv-2.4.13.7.tar.gz
-- Applying patch 0002-install-options.patch
-- Applying patch 0003-force-package-requirements.patch
-- Applying patch 0004-add-ffmpeg-missing-defines.patch
-- Applying patch 0005-fix-cuda.patch
-- Using source at F:/vcpkg_cenit/buildtrees/opencv2/src/2.4.13.7-64eb3208d2.clean
CMake Deprecation Warning at C:/Users/ssinigar/Codice/vcpkg_cenit/scripts/cmake/vcpkg_check_features.cmake:182 (message):
  calling `vcpkg_check_features` without the `FEATURES` keyword has been
  deprecated.

      Please add the `FEATURES` keyword to the call.
Call Stack (most recent call first):
  ports/opencv2/portfile.cmake:34 (vcpkg_check_features)
  C:/Users/ssinigar/Codice/vcpkg_cenit/scripts/ports.cmake:142 (include)


-- Found external ninja('1.10.2').
-- Configuring x64-windows
-- Building x64-windows-dbg
-- Building x64-windows-rel
-- Installing: F:/vcpkg_cenit/packages/opencv2_x64-windows/share/opencv2/copyright
-- Performing post-build validation
-- Performing post-build validation done
Uploading binaries for opencv2:x64-windows to NuGet source vcpkgbinarycache.
Building package opencv2[core,cuda,eigen,ffmpeg,jpeg,png,tiff]:x64-windows... done
Installing package opencv2[core,cuda,eigen,ffmpeg,jpeg,png,tiff]:x64-windows...
Installing package opencv2[core,cuda,eigen,ffmpeg,jpeg,png,tiff]:x64-windows... done
```

let me check it more, but again i cannot reproduce it locally

**Comment by NancyLi1013** [comment]
Can you also test `x64-windows-static` triplet?

**Comment by cenit** [comment]
x64-windows-static only opencv2[cuda] features or all of them?

**Comment by cenit** [comment]
`.\vcpkg.exe install darknet[cuda,opencv2-base,opencv2-cuda]:x64-windows-static --recurse`

```
Starting package 9/12: opencv2:x64-windows-static
Building package opencv2[core,cuda,eigen,ffmpeg,jpeg,png,tiff]:x64-windows-static...
-- Using cached F:/vcpkg_cenit/downloads/opencv-opencv-2.4.13.7.tar.gz
-- Cleaning sources at F:/vcpkg_cenit/buildtrees/opencv2/src/2.4.13.7-64eb3208d2.clean. Use --editable to skip cleaning for the packages you specify.
-- Extracting source F:/vcpkg_cenit/downloads/opencv-opencv-2.4.13.7.tar.gz
-- Applying patch 0002-install-options.patch
-- Applying patch 0003-force-package-requirements.patch
-- Applying patch 0004-add-ffmpeg-missing-defines.patch
-- Applying patch 0005-fix-cuda.patch
-- Using source at F:/vcpkg_cenit/buildtrees/opencv2/src/2.4.13.7-64eb3208d2.clean
CMake Deprecation Warning at C:/Users/ssinigar/Codice/vcpkg_cenit/scripts/cmake/vcpkg_check_features.cmake:182 (message):
  calling `vcpkg_check_features` without the `FEATURES` keyword has been
  deprecated.

      Please add the `FEATURES` keyword to the call.
Call Stack (most recent call first):
  ports/opencv2/portfile.cmake:34 (vcpkg_check_features)
  C:/Users/ssinigar/Codice/vcpkg_cenit/scripts/ports.cmake:142 (include)


-- Found external ninja('1.10.2').
-- Configuring x64-windows-static
-- Building x64-windows-static-dbg
-- Building x64-windows-static-rel
-- Installing: F:/vcpkg_cenit/packages/opencv2_x64-windows-static/share/opencv2/copyright
-- Performing post-build validation
-- Performing post-build validation done
Uploading binaries for opencv2:x64-windows-static to NuGet source vcpkgbinarycache.
Building package opencv2[core,cuda,eigen,ffmpeg,jpeg,png,tiff]:x64-windows-static... done
Installing package opencv2[core,cuda,eigen,ffmpeg,jpeg,png,tiff]:x64-windows-static...
Installing package opencv2[core,cuda,eigen,ffmpeg,jpeg,png,tiff]:x64-windows-static... done
Elapsed time for package opencv2:x64-windows-static: 39.13 min
...
Starting package 12/12: darknet:x64-windows-static
Building package darknet[core,cuda,opencv2-base,opencv2-cuda]:x64-windows-static...
-- Using cached F:/vcpkg_cenit/downloads/AlexeyAB-darknet-b6cd7592183043613f4b41c25e3385e7f752845c.tar.gz
-- Cleaning sources at F:/vcpkg_cenit/buildtrees/darknet/src/e7f752845c-b69c00dce8.clean. Use --editable to skip cleaning for the packages you specify.
-- Extracting source F:/vcpkg_cenit/downloads/AlexeyAB-darknet-b6cd7592183043613f4b41c25e3385e7f752845c.tar.gz
-- Using source at F:/vcpkg_cenit/buildtrees/darknet/src/e7f752845c-b69c00dce8.clean
-- Found external ninja('1.10.2').
-- Configuring x64-windows-static-dbg
-- Configuring x64-windows-static-rel
-- Building x64-windows-static-dbg
-- Building x64-windows-static-rel
-- Installing: F:/vcpkg_cenit/packages/darknet_x64-windows-static/tools/darknet/scripts/download_weights.ps1
-- To download weight files, please go to F:/vcpkg_cenit/installed/x64-windows-static/tools/darknet/scripts and run ./download_weights.ps1
-- Installing: F:/vcpkg_cenit/packages/darknet_x64-windows-static/share/darknet/copyright
-- Performing post-build validation
-- Performing post-build validation done
Uploading binaries for darknet:x64-windows-static to NuGet source vcpkgbinarycache.
Building package darknet[core,cuda,opencv2-base,opencv2-cuda]:x64-windows-static... done
Installing package darknet[core,cuda,opencv2-base,opencv2-cuda]:x64-windows-static...
Installing package darknet[core,cuda,opencv2-base,opencv2-cuda]:x64-windows-static... done
Elapsed time for package darknet:x64-windows-static: 2.888 min
```

**Comment by cenit** [comment]
opencv3-cuda and opencv-cuda [opencv4] depend on cudnn, which is not compatible with static triplets even if manually installed

**Comment by cenit** [comment]
@strega-nil the manifest file vcpkg.json in this port is identical to the one published in the upstream repo: https://github.com/AlexeyAB/darknet/blob/master/vcpkg.json

is there any way to use directly the upstream manifest instead of having to copy it here? or is it a feature that i'm asking that is already planned for the future? or is it considered a wrong idea somehow?

**Review by NancyLi1013 — APPROVED** [review]

**Comment by cenit** [comment]
> @cenit We will make this PR high priority.

any news? please...

**Comment by JackBoosY** [comment]
Ping @strega-nil @ras0219-msft for review and merge this PR first, it lasts so long.

**Comment by cenit** [comment]
> Ping @strega-nil @ras0219-msft for review and merge this PR first, it lasts so long.

😢 🤞

**Comment by BillyONeal** [comment]
Thanks for your contribution! (And sorry for the runaround)

**Pull request merged**